### PR TITLE
Add test for ECP group metadata

### LINKED ISSUE #5919: Test ECP group metadata
For each supported ECP group (values of `mbedtls_ecp_group_id`), have a unit test that calls does the following:

* Call `mbedtls_ecp_group_load`.
* Check the values of the numbers in the group structure. Copy the test data from a reference (e.g. SEC2), not from the code. Must catch https://github.com/Mbed-TLS/mbedtls/pull/5810.
* Call `mbedtls_ecp_curve_info_from_grp_id` and check the values that it returns. Call `mbedtls_ecp_curve_info_from_tls_id` and `mbedtls_ecp_curve_info_from_name` on the `tls_id` and `name` in the curve info and check that you get the same curve info.
* Call `mbedtls_ecp_tls_write_group` and validate the output against test data. Call `mbedtls_ecp_tls_read_group_id` and `mbedtls_ecp_tls_read_group` on the output and check you get back the original group ID.
* Check `mbedtls_ecp_get_type`, `mbedtls_ecdh_can_do` and `mbedtls_ecdsa_can_do`.
* Call `mbedtls_ecp_group_copy` on the group.
* Check that the group appears in `mbedtls_ecp_curve_list()` and in `mbedtls_ecp_grp_id_list()`.

Related: https://github.com/Mbed-TLS/mbedtls/issues/5920

### PR DESCRIPTION
**Pull request by wernerlewis**
## Description
Test cases added to check curve parameters and information for each
curve supported. Expected parameters are taken from references: SEC 2
for NIST, RFC 5639 for Brainpool, and RFC 7748 for curve25519/curve448.

Each group is compared with known curve parameters, other metadata is
checked and validated against test data. This includes loading the group by
TLS ID, by name, and from the result of `mbedtls_ecp_tls_write_group`.

This has been tested with the bug described in #5810, which is caught by
these tests. Resolves #5919.

## Status
**READY**

## Requires Backporting
Yes 2.28 #6204

### CODE REVIEW

File: tests/suites/test_suite_ecp.data
Comment (daverodgman) [inline tests/suites/test_suite_ecp.data:362]: This seems correct but maybe a bit confusing, in that the value for A (according to RFC7748) is actually 156326, but Mbed TLS uses (A+2)/4 rather than A itself. Similarly for Curve25519 above.

Ideally it would be nice if the numbers in the test data matched the spec exactly, and test_suite_ecp.function converts them to Mbed TLS' internal representation before checking (using `crv_type` to determine if this is needed).

**Review by daverodgman — APPROVED** [review]

File: tests/suites/test_suite_ecp.data
Comment (AndrzejKurek) [inline tests/suites/test_suite_ecp.data:353]: This should be `MBEDTLS_ECP_DP_BP512R1_ENABLED` instead. Interesting that this wasn't picked up by scripts as a typo...

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
I verified that all of https://github.com/Mbed-TLS/mbedtls/issues/5919 apart from one (minor) have been addressed:
- `mbedtls_ecp_tls_read_group_id` is not called explicitly, but is called via `mbedtls_ecp_tls_read_group`.
I also verified that all of the data from RFC's is copied properly, apart from one detail: leading zeroes are removed. The library should handle it fine and I would even consider that a part of being able to handle raw data from RFC's, so it would be great to fix that. Affected curves:
- secp224k1;
- secp521r1;
- brainpoolP384r1.
One major typo found, leading zeroes are a minor and just a reccomendation from me.

File: tests/suites/test_suite_ecp.data
Comment (AndrzejKurek) [reply tests/suites/test_suite_ecp.data:353]: Right, `check_names.py` doesn't check tests.

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_ecp.data]: We can find out about typos in test dependencies from the result analysis job on Jenkins: it reports test cases that are not executed in any configuration. It isn't a CI failure to have those because some are legitimate and [we haven't gotten around to analyzing the logs](https://github.com/Mbed-TLS/mbedtls/issues/5390).

**Review by daverodgman — APPROVED** [review]

File: tests/suites/test_suite_ecp.function
Comment (AndrzejKurek) [inline tests/suites/test_suite_ecp.function:875]: `mbedtls_ecp_tls_read_group_id` will modify `vbuf`.
```suggestion
    TEST_EQUAL( read_g_id, id );
    vbuf = buf;
```

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
One small change that fails on the CI.

**Review by AndrzejKurek — APPROVED** [review]

**Review by daverodgman — APPROVED** [review]

**Comment by mpg** [comment]
@daverodgman Why was this merged while it was labeled needs-backport? As a general rule we wait for backports to be ready before we merge the main PR, and it's OK to make exceptions but IMO we should briefly state the reason why before merging, at least to make it clear that it was a conscious decision.

**Comment by mpg** [comment]
Note: I just merged the backport https://github.com/Mbed-TLS/mbedtls/pull/6204 (this is how I noticed) and I don't think it's a problem that it won't be in the same release as the main PR, considering this is just a test enhancement.

**Pull request merged**
